### ISSUE TITLE
Non-fatal `libGL error: failed to load driver` errors are thrown during `sct_check_dependencies`

### ISSUE THREAD

**Issue by joshuacwnewton** [spinalcordtoolbox/spinalcordtoolbox#3927]
Originally reported by @mguaypaq in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3915#pullrequestreview-1157166675:

> Just today I noticed this chattiness of the matplotlib test on my laptop:
> 
> ```
> Check PropSeg compatibility with OS ................[OK]
> Check if figure can be opened with matplotlib.......Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
> libGL error: MESA-LOADER: failed to open iris: /usr/lib/dri/iris_dri.so: cannot open shared object file: No such file or directory (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri, suffix _dri)
> libGL error: failed to load driver: iris
> libGL error: MESA-LOADER: failed to open swrast: /usr/lib/dri/swrast_dri.so: cannot open shared object file: No such file or directory (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri, suffix _dri)
> libGL error: failed to load driver: swrast
> [OK] (Using GUI backend: 'QtAgg')
> Check if figure can be opened with PyQt.............[OK]
> ```

**Comment by joshuacwnewton** [comment]
Note that there are two checks here for both `matplotlib` and `PyQt`. And, in PR https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3916, the order of these two checks were swapped.

After the swap, the errors now occur at the `PyQt` step instead of the `matplotlib` step: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3925#issuecomment-1297354520

So, it seems like this isn't necessarily a `matplotlib` issue.

**Comment by joshuacwnewton** [comment]
tl;dr, I was able to remove the warnings using two steps:

```bash
# 1. Softlink missing dri shared object files to expected directory
ln -s /usr/lib/x86_64-linux-gnu/dri/ /usr/lib/dri

# 2. Remove outdated `libstdc++.so` file that comes with SCT's venv
mv $SCT_DIR/python/envs/venv_sct/lib/libstdc++.so.6.0.29 $SCT_DIR/python/envs/venv_sct/lib/libstdc++.so.6.0.29.bak
```

<details>
<summary> Detailed investigation </summary>

### 1. Softlinking the correct `dri` directory

> ```
> libGL error: MESA-LOADER: failed to open iris: /usr/lib/dri/iris_dri.so: cannot open shared object file: No such file or directory (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri, suffix _dri)
> ```

I noticed that `MESA-LOADER` is looking in `/usr/lib/dri/` for the shared object files. But, on my system, this directory doesn't exist, and instead these shared object files are actually located in `/usr/lib/x86_64-linux-gnu/dri`.

If I create a softlink from `/usr/lib/x86_64-linux-gnu/dri` to `/usr/lib/dri/`, the errors are silenced, but new errors appear.

### 2. Addressing "version `GLIBCXX_3.4.30' not found" error

The new errors are as follows:

```console
joshua@tadpole:~/repos/spinalcordtoolbox$ sct_check_dependencies
[...]
Check if figure can be opened with PyQt.............libGL error: MESA-LOADER: failed to open iris: /home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /lib/x86_64-linux-gnu/libLLVM-14.so.1) (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri, suffix _dri)
libGL error: failed to load driver: iris
libGL error: MESA-LOADER: failed to open iris: /home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /lib/x86_64-linux-gnu/libLLVM-14.so.1) (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri, suffix _dri)
libGL error: failed to load driver: iris
libGL error: MESA-LOADER: failed to open swrast: /home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /lib/x86_64-linux-gnu/libLLVM-14.so.1) (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri, suffix _dri)
libGL error: failed to load driver: swrast
[OK]
```

<details>
<summary> LIBGL_DEBUG=verbose </summary>

```console
joshua@tadpole:~/repos/spinalcordtoolbox$ LIBGL_DEBUG=verbose sct_check_dependencies
[...]
Check if figure can be opened with PyQt.............libGL: Can't open configuration file /etc/drirc: No such file or directory.
libGL: Can't open configuration file /home/joshua/.drirc: No such file or directory.
libGL: using driver i915 for 18
libGL: Can't open configuration file /etc/drirc: No such file or directory.
libGL: Can't open configuration file /home/joshua/.drirc: No such file or directory.
libGL: using driver i915 for 18
libGL: pci id for fd 18: 8086:9a78, driver iris
libGL: MESA-LOADER: failed to open /usr/lib/x86_64-linux-gnu/dri/iris_dri.so: /home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /lib/x86_64-linux-gnu/libLLVM-14.so.1)
libGL: MESA-LOADER: failed to open \$${ORIGIN}/dri/iris_dri.so: \$${ORIGIN}/dri/iris_dri.so: cannot open shared object file: No such file or directory
libGL: MESA-LOADER: failed to open /usr/lib/dri/iris_dri.so: /home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /lib/x86_64-linux-gnu/libLLVM-14.so.1)
libGL error: MESA-LOADER: failed to open iris: /home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /lib/x86_64-linux-gnu/libLLVM-14.so.1) (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri, suffix _dri)
libGL error: failed to load driver: iris
libGL: using driver i915 for 18
libGL: Can't open configuration file /etc/drirc: No such file or directory.
libGL: Can't open configuration file /home/joshua/.drirc: No such file or directory.
libGL: using driver i915 for 18
libGL: pci id for fd 18: 8086:9a78, driver iris
libGL: MESA-LOADER: failed to open /usr/lib/x86_64-linux-gnu/dri/iris_dri.so: /home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /lib/x86_64-linux-gnu/libLLVM-14.so.1)
libGL: MESA-LOADER: failed to open \$${ORIGIN}/dri/iris_dri.so: \$${ORIGIN}/dri/iris_dri.so: cannot open shared object file: No such file or directory
libGL: MESA-LOADER: failed to open /usr/lib/dri/iris_dri.so: /home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /lib/x86_64-linux-gnu/libLLVM-14.so.1)
libGL error: MESA-LOADER: failed to open iris: /home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /lib/x86_64-linux-gnu/libLLVM-14.so.1) (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri, suffix _dri)
libGL error: failed to load driver: iris
libGL: MESA-LOADER: failed to open /usr/lib/x86_64-linux-gnu/dri/swrast_dri.so: /home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /lib/x86_64-linux-gnu/libLLVM-14.so.1)
libGL: MESA-LOADER: failed to open \$${ORIGIN}/dri/swrast_dri.so: \$${ORIGIN}/dri/swrast_dri.so: cannot open shared object file: No such file or directory
libGL: MESA-LOADER: failed to open /usr/lib/dri/swrast_dri.so: /home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /lib/x86_64-linux-gnu/libLLVM-14.so.1)
libGL error: MESA-LOADER: failed to open swrast: /home/joshua/repos/spinalcordtoolbox/python/envs/venv_sct/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /lib/x86_64-linux-gnu/libLLVM-14.so.1) (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri, suffix _dri)
libGL error: failed to load driver: swrast
[OK]
```
</details>


I then noticed that the `libstdc++.so.6` file in SCT's virtualenv is v29:

```console
joshua@tadpole:~/repos/spinalcordtoolbox$ ls -l python/envs/venv_sct/lib | grep libstdc++.so
lrwxrwxrwx  1 joshua joshua       19 Nov  1 10:10 libstdc++.so -> libstdc++.so.6.0.29
lrwxrwxrwx  1 joshua joshua       19 Nov  1 10:10 libstdc++.so.6 -> libstdc++.so.6.0.29
-rwxrwxr-x  2 joshua joshua 17981480 Nov  1 10:10 libstdc++.so.6.0.29.bak
```

Whereas my system reports having the proper v30:

```console
joshua@tadpole:~/repos/spinalcordtoolbox$ strings /usr/lib/x86_64-linux-gnu/libstdc++.so.6 | grep GLIBCXX
GLIBCXX_3.4
GLIBCXX_3.4.1
GLIBCXX_3.4.2
[...]
GLIBCXX_3.4.28
GLIBCXX_3.4.29
GLIBCXX_3.4.30
```

Which makes me think we're running into:

* https://github.com/ContinuumIO/anaconda-issues/issues/13013#issuecomment-1224401842

> When I try `strings /usr/lib/x86_64-linux-gnu/libstdc++.so.6 | grep GLIBCXX` I see that I have up to and including `GLIBCXX_3.4.30`. 

> A workaround that I'm not sure is entirely going to work well in the anaconda environment is to just remove the `libstdc++.so.6` from the anaconda library. I ran `mv libstdc++.so.6.0.28 libstdc++.so.6.0.28.bak` so executables wouldn't find the anaconda library anymore, and now everything seems to work fine just linking to the system installed libraries.

Following the advice from https://github.com/ContinuumIO/anaconda-issues/issues/13013#issuecomment-1224444761, I ran:

```bash
mv $SCT_DIR/python/envs/venv_sct/lib/libstdc++.so.6.0.29 $SCT_DIR/python/envs/venv_sct/lib/libstdc++.so.6.0.29.bak
```

This completely removes the warning.

----

</details>

I don't feel entirely confident in these steps, though. I'd like to answer the following questions:

* Why is `MESA-LOADER` looking in the `/usr/lib/dri` folder?
* Why have the `dri` files moved into the `x86_64-linux-gnu` subfolder?
* Is there a way to change the search path using some sort of environment variable (rather than softlinking the directory)?
* Is there a way to include the most-up-to-date `libstdc++.so.6` file inside the `venv_sct` conda environment?

**Comment by joshuacwnewton** [comment]
> Is there a way to include the most-up-to-date `libstdc++.so.6` file inside the `venv_sct` conda environment?

ANS: Yes! From [this SO thread]():

> First of all, make sure that you have the latest lib with `conda install -c conda-forge libstdcxx-ng`

Running this command in the active `venv_sct` results in the correct `libstdc++.so.6.0.30` file being present in `venv_sct/lib`.

**Comment by joshuacwnewton** [comment]
I tried deleting the `dri/` symlink (only relying on the `conda install -c conda-forge libstdcxx-ng` fix) and it seems to eliminate the error entirely!

**Comment by joshuacwnewton** [comment]
I'm encountering these error messages once more in my Debian 11 dev environment. (Context: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3367.)

**Comment by joshuacwnewton** [comment]
I tried rerunning `sct_check_dependencies`, and now I'm encountering log messages with a different kind of error:

```
Check if figure can be opened with PyQt.............libGL error: MESA-LOADER: failed to open iris: /lib/x86_64-linux-gnu/libLLVM-11.so.1: undefined symbol: ffi_type_sint32, version LIBFFI_BASE_7.0 (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
libGL error: failed to load driver: iris
libGL error: MESA-LOADER: failed to open iris: /lib/x86_64-linux-gnu/libLLVM-11.so.1: undefined symbol: ffi_type_sint32, version LIBFFI_BASE_7.0 (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
libGL error: failed to load driver: iris
libGL error: MESA-LOADER: failed to open swrast: /lib/x86_64-linux-gnu/libLLVM-11.so.1: undefined symbol: ffi_type_sint32, version LIBFFI_BASE_7.0 (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
libGL error: failed to load driver: swrast
```

**Comment by joshuacwnewton** [comment]
I'm reopening this issue because I think it might be good to considering adding `conda install libffi -c conda-forge` to `install_sct`. That way, we avoid this warnings on fresh installations.

**Comment by joshuacwnewton** [comment]
> (I also found that `conda-forge`'s copy of `libffi`, which is `3.4.2`, doesn't include the 7 -> 8 symlinks at all, which _also_ solves the problem (due to the fallback to system copies of libffi.))

I've reopened this issue because I think it may be good to add `conda install libffi -c conda-forge` .

If we do this, there is a risk that the `conda-forge` version of `libffi` may update to `3.4.4`, and the warnings may be suddenly introduced within a future stable release. 

To try and mitigate this, though, I have opened an issue in the conda-forge feedstock repo so that they are aware:

- https://github.com/conda-forge/libffi-feedstock/issues/46

**Comment by joshuacwnewton** [comment]
> If we do this, there is a risk that the `conda-forge` version of `libffi` may update to `3.4.4`, and the warnings may be suddenly introduced within a future stable release.

The conda-forge libffi maintainer has confirmed that this is a [`defaults`-only issue](https://github.com/conda-forge/libffi-feedstock/issues/46#issuecomment-1607751709), and so I'm now much more confident that sticking with `conda-forge` is a safe thing to do long-term.

**Comment by aopolin-lv** [comment]
> Separately from the aforementioned "`No such file or directory`" issues:
> 
> I tried reinstalling SCT on my Debian 11 dev machine, and now I'm encountering similar log messages, but with a different kind of error:
> 
> ```
> Check if figure can be opened with PyQt.............libGL error: MESA-LOADER: failed to open iris: /lib/x86_64-linux-gnu/libLLVM-11.so.1: undefined symbol: ffi_type_sint32, version LIBFFI_BASE_7.0 (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
> libGL error: failed to load driver: iris
> libGL error: MESA-LOADER: failed to open iris: /lib/x86_64-linux-gnu/libLLVM-11.so.1: undefined symbol: ffi_type_sint32, version LIBFFI_BASE_7.0 (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
> libGL error: failed to load driver: iris
> libGL error: MESA-LOADER: failed to open swrast: /lib/x86_64-linux-gnu/libLLVM-11.so.1: undefined symbol: ffi_type_sint32, version LIBFFI_BASE_7.0 (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
> libGL error: failed to load driver: swrast
> ```
> 
> **EDIT:**
> 
> From what I can tell, this libffi-based issue is caused by a recent update to conda's `libffi` package (`libffi-3.4.4`). Looking at the `lib` folder in the conda environment for a fresh install of SCT, we see:
> 
> ```
> (venv_sct) master|joshua@tadpole:~/repos/spinalcordtoolbox$ ls -lah python/envs/venv_sct/lib | grep ffi
> lrwxrwxrwx  1 joshua joshua   15 Jun  2 12:26 libffi.7.so -> libffi.so.8.1.2
> lrwxrwxrwx  1 joshua joshua   15 Jun  2 12:26 libffi.8.so -> libffi.so.8.1.2
> -rw-rw-r--  2 joshua joshua  96K Jun  2 12:26 libffi.a
> lrwxrwxrwx  1 joshua joshua   15 Jun  2 12:26 libffi.so -> libffi.so.8.1.2
> lrwxrwxrwx  1 joshua joshua   15 Jun  2 12:26 libffi.so.7 -> libffi.so.8.1.2
> lrwxrwxrwx  1 joshua joshua   15 Jun  2 12:26 libffi.so.8 -> libffi.so.8.1.2
> -rwxrwxr-x  2 joshua joshua  71K Jun  2 12:26 libffi.so.8.1.2
> ```
> 
> Here,`libffi.so.7` is linked to `libffi.so.8.1.2`! When comparing that to my local copy of libffi7, I see:
> 
> ```
> joshua@tadpole:~$ ls -lah /usr/lib/x86_64-linux-gnu/ | grep libffi
> -rw-r--r--   1 root   root     60K Feb 20  2021 libffi.a
> -rw-r--r--   1 root   root     60K Feb 20  2021 libffi_pic.a
> lrwxrwxrwx   1 root   root      15 Feb 20  2021 libffi.so -> libffi.so.7.1.0
> lrwxrwxrwx   1 root   root      15 Feb 20  2021 libffi.so.7 -> libffi.so.7.1.0
> -rw-r--r--   1 root   root     43K Feb 20  2021 libffi.so.7.1.0
> ```
> 
> Given that the error message suggests that libLLVM is looking for a specific symbol from `LIBFFI_BASE_7.0`, I tried temporarily removing the symlinks in SCT's conda environment, so that libLLVM would fallback to the system `libffi7` files.
> 
> ```
> mv python/envs/venv_sct/lib/libffi.7.so python/envs/venv_sct/lib/libffi.7.so.bak
> mv python/envs/venv_sct/lib/libffi.so.7 python/envs/venv_sct/lib/libffi.so.7.bak
> ```
> 
> And, the warning goes away entirely.
> 
> (I also found that `conda-forge`'s copy of `libffi`, which is `3.4.2`, doesn't include the 7 -> 8 symlinks at all, which _also_ solves the problem (due to the fallback to system copies of libffi.))
> 
> So, it seems this issue was introduced in either 3.4.3 or 3.4.4 of conda's libffi package. Thus, I've reported the issue upstream here: [ContinuumIO/anaconda-issues#13205](https://github.com/ContinuumIO/anaconda-issues/issues/13205)
> 
> tl;dr, my best understanding of my problem here is that:
> 
> * I'm on Debian 11, which uses an _old_ version of libLLVM (11, newest stable is 15), which is turn relies on a version of libffi with an older ABI (ABI 7)
> * Conda's `libffi` package has upgraded from ABI 7 to ABI 8, symlinking 7 -> 8 for compatibility.
> * ABI 8 isn't compatible with libLLVM11, however, causing the symbol errors.
> * So, the solution is to make sure that `libffi.so.7` is properly linked to an ABI 7-based file (or, presumably on non-Debian systems, an upgraded libLLVM solves this issue too?)
> 
> I'm not sure if this is relevant for other Debian systems? (Our CI won't catch this because it's headless, and thus can't test the code that would trigger this issue.)
> 
> But, hopefully this write-up saves us some headache in the future if/when a Debian user tries to install SCT and encounters the same issue.

It's so cool. I was stuck in this problem for nearly two days, and I didn't solve it until I saw your comments.

**Comment by joshuacwnewton** [comment]
I'm so glad to hear it! You may wish to add a comment to the following upstream issue:

- https://github.com/AnacondaRecipes/libffi-feedstock/issues/15

That way, the maintainers of `libffi` know that there are other people who are struggling with this issue.
